pkg/util/kernel/proc.go

```diff
@@ -56,6 +56,141 @@ func WithAllProcs(procRoot string, fn func(int) error) error {
 	return nil
 }
 
+// ProcMapEntry represents an entry of the /proc/X/maps file
+type ProcMapEntry struct {
+	Start   uint64
+	End     uint64
+	Mode    string
+	Offset  uint64
+	MajorID uint64
+	MinorID uint64
+	Inode   uint64
+	Path    string
+	Deleted bool
+}
+
+// ProcMapEntries is a collection of ProcMapEntry objects sorted by start address
+type ProcMapEntries []ProcMapEntry
+
+// Insert will insert the entry in the correct position in the slice based on its start address
+func (p *ProcMapEntries) Insert(entry ProcMapEntry) {
+	for i, other := range *p {
+		if entry.Start < other.Start {
+			if i == 0 {
+				*p = append([]ProcMapEntry{entry}, *p...)
+				return
+			}
+
+			*p = append(append((*p)[:i-1], entry), (*p)[i:]...)
+			return
+		}
+	}
+
+	*p = append(*p, entry)
+}
+
+// FindEntryForAddress will return the entry that contains the given address
+func (p *ProcMapEntries) FindEntryForAddress(addr uint64) *ProcMapEntry {
+	// Probably could be made more efficient, but for a first implementation it should be fine
+	for i := range *p {
+		if addr >= (*p)[i].Start && addr < (*p)[i].End {
+			return &(*p)[i]
+		}
+	}
+
+	return nil
+}
+
+// readProcessMemMapsFromBuffer reads the content of a /proc/PID/maps file from a buffer
+// Format reference: https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/fs/proc/task_mmu.c#n304
+// see show_vma_header_prefix function
+func readProcessMemMapsFromBuffer(buffer io.Reader) (ProcMapEntries, error) {
+	scanner := bufio.NewScanner(buffer)
```

Comment:
i saw a similar [implementation](https://github.com/DataDog/datadog-agent/blob/d75437a2f076228280d2cc2d28a0535c285be446/pkg/network/usm/sharedlibraries/watcher.go#L116) in USM 

it has some differences in the actual logic of parsing the lines of the maps file.

1) do we need to have a similar cache for the edge case where the same file is loaded more than once?
2) i saw that their line structure is much more strict than yours. do you know to explain the differences?

Reply:
Marking this as resolved as we're using prometheus/procfs for the parsing.